crates/ra_assists/src/handlers/replace_qualified_name_with_use.rs

```diff
@@ -39,16 +43,28 @@ pub(crate) fn replace_qualified_name_with_use(
         target,
         |builder| {
             let path_to_import = hir_path.mod_path().clone();
+            let container = match find_insert_use_container(path.syntax(), ctx) {
+                Some(c) => c,
+                None => return,
+            };
             insert_use_statement(path.syntax(), &path_to_import, ctx, builder.text_edit_builder());
 
-            if let Some(last) = path.segment() {
-                // Here we are assuming the assist will provide a correct use statement
-                // so we can delete the path qualifier
-                builder.delete(TextRange::new(
-                    path.syntax().text_range().start(),
-                    last.syntax().text_range().start(),
-                ));
+            // Now that we've brought the name into scope, re-qualify all paths that could be
+            // affected (that is, all paths inside the node we added the `use` to).
+            let hir_path = match hir::Path::from_ast(path.clone()) {
+                Some(p) => p,
+                None => return,
+            };
+            let mut rewriter = SyntaxRewriter::default();
+            match container {
```

Comment:
I belive there's a combinator to map `Either<A, B>` to `T` which you can use to avoid duplicate arms.